src/kernel/faulthandler.c

```diff
@@ -43,129 +30,78 @@ bool_t sendFaultIPC(tcb_t *tptr, cap_t handlerCap, bool_t can_donate)
-    }
+    cap_t handlerCap = TCB_PTR_CTE_PTR(tptr, tcbTimeoutHandler)->cap;
+    (void)sendFaultIPC(tptr, handlerCap, false, fault);
 }
 
-exception_t sendFaultIPC(tcb_t *tptr)
+#endif /* CONFIG_KERNEL_MCS */
+
+void handleFault(tcb_t *tptr)
 {
-    cptr_t handlerCPtr;
-    cap_t  handlerCap;
-    lookupCap_ret_t lu_ret;
-    lookup_fault_t original_lookup_fault;
-
-    original_lookup_fault = current_lookup_fault;
-
-    handlerCPtr = tptr->tcbFaultHandler;
-    lu_ret = lookupCap(tptr, handlerCPtr);
-    if (lu_ret.status != EXCEPTION_NONE) {
-        current_fault = seL4_Fault_CapFault_new(handlerCPtr, false);
-        return EXCEPTION_FAULT;
+    seL4_Fault_t fault = current_fault;
+
+#ifdef CONFIG_KERNEL_MCS
+
+    if (sendFaultIPC(tptr,
+                     TCB_PTR_CTE_PTR(tptr, tcbFaultHandler)->cap,
+                     (tptr->tcbSchedContext != NULL),
+                     fault)) {
+        return; /* fault handler will take care of the fault */
     }
-    handlerCap = lu_ret.cap;
-
-    if (cap_get_capType(handlerCap) == cap_endpoint_cap &&
-        cap_endpoint_cap_get_capCanSend(handlerCap) &&
-        (cap_endpoint_cap_get_capCanGrant(handlerCap) ||
-         cap_endpoint_cap_get_capCanGrantReply(handlerCap))) {
-        tptr->tcbFault = current_fault;
-        if (seL4_Fault_get_seL4_FaultType(current_fault) == seL4_Fault_CapFault) {
-            tptr->tcbLookupFailure = original_lookup_fault;
+
+#else /* not CONFIG_KERNEL_MCS */
+
+    lookup_fault_t original_lookup_fault = current_lookup_fault;
+    cptr_t handlerCPtr = tptr->tcbFaultHandler;
+    lookupCap_ret_t lu_ret = lookupCap(tptr, handlerCPtr);
+    exception_t status = lu_ret.status;
+    if (EXCEPTION_NONE == status) {
+        cap_t handlerCap = lu_ret.cap;
+        if (cap_get_capType(handlerCap) == cap_endpoint_cap &&
+            cap_endpoint_cap_get_capCanSend(handlerCap) &&
+            (cap_endpoint_cap_get_capCanGrant(handlerCap) ||
+             cap_endpoint_cap_get_capCanGrantReply(handlerCap))) {
+            tptr->tcbFault = fault;
+            if (seL4_Fault_get_seL4_FaultType(fault) == seL4_Fault_CapFault) {
+                tptr->tcbLookupFailure = original_lookup_fault;
+            }
+            sendIPC(true, true,
+                    cap_endpoint_cap_get_capEPBadge(handlerCap),
+                    cap_endpoint_cap_get_capCanGrant(handlerCap), true, tptr,
+                    EP_PTR(cap_endpoint_cap_get_capEPPtr(handlerCap)));
+            return; /* fault handler will take care of the fault */
         }
-        sendIPC(true, true,
-                cap_endpoint_cap_get_capEPBadge(handlerCap),
-                cap_endpoint_cap_get_capCanGrant(handlerCap), true, tptr,
-                EP_PTR(cap_endpoint_cap_get_capEPPtr(handlerCap)));
 
-        return EXCEPTION_NONE;
-    } else {
-        current_fault = seL4_Fault_CapFault_new(handlerCPtr, false);
         current_lookup_fault = lookup_fault_missing_capability_new(0);
```

Comment:
I wonder, do we really have to update `current_lookup_fault` here? Seems no body uses this.